### DISCUSSION TITLE
API Accestoken validation slow

### DISCUSSION THREAD

**Discussion by cocadepoma** [zitadel/zitadel#7172] [Q&A] [answered]
Hello, I am trying [ZITADEL's quote example app](https://github.com/zitadel/example-quote-generator-app/tree/main) and I notice that the '/api/custom_quote' endpoint takes around 0.4-0.5s to response.

I would like to know if this is normal or if there is a way to validate the tokens without calling ZITADEL's API and how to do it in Python.

I read that a JWT with Private Key can be used here, but how to use it is unclear.

Thanks for your help.

**Comment by dakshitha** [comment]
Hi! The example calls the ZITADEL introspection endpoint and to reduce the latency, you can use JWKs(JSON Web Keys). 

You can cache the JWKs obtained from ZITADEL's endpoint (/oauth/v2/keys) to validate the JWTs. ZITADEL updates its key pairs every 6 hours and publishes new JWKs with unique kid (key IDs). Caching allows you to keep up with these updates without frequent API calls.

Here’s how you can implement it: 
Store the keys in a local cache after fetching them from ZITADEL’s /oauth/v2/keys endpoint.

When you need to validate a JWT, first check if the relevant key (based on the kid in the JWT header) is in your cache. If it’s not, fetch the latest keyset from ZITADEL, cache it, and make a note of the fetch to avoid redundant calls in case of invalid kid values being presented.

Libraries like pyjwt are handy for dealing with JWTs in Python. They allow you to validate the JWT’s signature using the cached JWK to ensure the token’s authenticity.

**Reply by cocadepoma** [reply]
Thank you, I think I understand what you're telling me, but I can't seem to get it to work correctly. I'm a bit lost.

Ignoring the caching issue, is this the way it should work?

```python
def get_fake_public_key():
    return {
      "use": "sig",
      "kty": "RSA",
      "kid": "1234",
      "alg": "RS256",
      "n": "rasfasFASFASF213",
      "e": "AQAB"
    }

public_key = get_fake_public_key()

decoded_token = jwt.decode(token_string, key=public_key, algorithms=["RS256"])
```

Because with that code, I get the message 'Invalid access token: Not enough segments' because the accessToken is not a valid JWT. What am I doing wrong?

**Reply by dakshitha** [reply, marked as answer]
Regarding the error, the problem is with the generated access token and my hunch is that if you followed the tutorial, you generated an opaque token (Bearer Token as shown in the UI) for the front-end app and not a JWT. Select JWT instead and generate the token again. 

<img width="653" alt="image" src="https://github.com/zitadel/zitadel/assets/11885043/c07f33e8-f3b7-4135-916f-8f177209088b">

Your code implementation looks correct. Make sure that you properly extract the public key for the JWKS based on the kid in the JWT header. Also, including the `aud` and `iss` claims in your validation process provides additional security and context, which is generally recommended.

e.g., 

```python
jwks_url = 'https://your-zitadel-domain.com/oauth/v2/keys'
jwks_response = requests.get(jwks_url)
jwks_data = jwks_response.json()

def get_public_key(kid):
    for key in jwks_data['keys']:
        if key['kid'] == kid:
            return RSAAlgorithm.from_jwk(key)
    return None

# Your JWT access token
token_string = 'your-access-token-here'
decoded_token = jwt.decode(
    token_string,
    key=get_public_key(decoded_token['kid']),  # Use the key associated with the token's kid
    algorithms=['RS256'],
    audience='your-client-id',  # Replace with your client ID
    issuer='https://your-zitadel-domain.com/',  # Replace with your ZITADEL domain
)

print(decoded_token)
```

**Reply by cocadepoma** [reply]
Hi  @dakshitha, it worked like a charm!
I did not find in the documentation how to do this, it would be nice for others to have this example, I owe you a coffee, thanks!

**Reply by dakshitha** [reply]
Super! Glad that it worked :)! Will include a JWKS example in our docs.

**Reply by davidhuser** [reply]
is there already a JWKS example in the docs @dakshitha ? Thank you

**Reply by dakshitha** [reply]
No, sorry. Couldn't add it yet (I will get to it soon),  but it's mostly around what was discussed here. Are you stuck somewhere?
